src/lib/lib.jl

```diff
@@ -112,6 +112,14 @@ end
 @adjoint getindex(xs::NTuple{N,Any}, r::AbstractUnitRange) where N =
   (xs[r], Δ -> (ntuple(j -> j in r ? Δ[findfirst(isequal(j), r)] : nothing, Val(N)), nothing))
 
+@adjoint function getindex(xs::NTuple{N,Any}, r::AbstractVector) where N
+  val = xs[r]
+  function back(Δ)
+    (ntuple(x -> x in r ? count(isequal(x), r) * Δ[findfirst(isequal(x), r)] : nothing, Val(N)), nothing)
```

Comment:
Faster, and IMO clearer:

```suggestion
    x̄s = ntuple(Val(length(xs))) do x
        total = zero(eltype(Δ))
        for r_i in eachindex(r)
            r[r_i] === x || continue
            total += Δ[r_i]
        end
        return total
    end
    return (x̄s, nothing)
```

Benchmarks:
```julia
function back_orig(xs, r)
    Δ = ones(length(xs[r]))  # for testing
    (ntuple(x -> x in r ? count(isequal(x), r) * Δ[findfirst(isequal(x), r)] : nothing, Val(length(xs))), nothing)
end

function back_neo(xs, r)
    Δ = ones(length(xs[r]))  # for testing
    dxs = ntuple(Val(length(xs))) do x
        total = zero(eltype(Δ))
        for r_i in eachindex(r)
            r[r_i] === x || continue
            total += Δ[r_i]
        end
        return total
    end
    return (dxs, nothing)
end


const xs1 = (10, 20, 30, 40, 50);
const r1 = [1,1,2,1,3,2,3,4,5,1];
```

```julia
julia> @btime back_neo(xs1, r1)
  391.060 ns (3 allocations: 416 bytes)
((4.0, 2.0, 2.0, 1.0, 1.0), nothing)

julia> @btime back_orig(xs1, r1)
  591.661 ns (10 allocations: 592 bytes)
((4.0, 2.0, 2.0, 1.0, 1.0), nothing)
```
```

Reply:
Mostly I wrote the above to convince myself that what this code was doing even made sense.
Does it have a test?

Reply:
Should be a function. The existing tests might be plenty in this case. I'm happy to take this suggestion too. It's a bit more verbose but I think that's alright.

Reply:
The suggestion is for a function.
I just didn't make it a function for the benchmark